### PR TITLE
Implement CanvasRenderingContext2d.fillText

### LINKED ISSUE #11681: Implement CanvasRenderingContext2d.fillText
Spec: https://html.spec.whatwg.org/multipage/scripting.html#dom-context-2d-filltext
Code: `components/script/dom/canvasrenderingcontext2d.rs`, `components/script/dom/weibidls/CanvasRenderingContext2D.webidl`, `components/canvas/canvas_paint_thread.rs`

(I hope I identified correctly the code files - followed the style in #5411)

### PR DESCRIPTION
**Pull request by utsavoza**
The PR consists of broadly two main changes:
- Implementation of Canvas2dRenderingContext.font
- Basic implementation of Canvas2dRenderingContext.fillText

Although I am not fully sure about the long term goals for the canvas backend in Servo, I assumed limited scope for font and text handling (should support simple text drawing with font selection) in the current implementation as I believe a more complete implementation would eventually be brought in as a part of #22957.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #11681
- [x] There are tests for these changes

### CODE REVIEW

**Comment by utsavoza** [comment]
@jdm, under what circumstances can [FontTemplateData::bytes](https://doc.servo.org/gfx/platform/freetype/font_template/struct.FontTemplateData.html#method.bytes) panic with message `No URL for Core Text font!`? I am unfamiliar with the underlying `core-text` implementation, but I am persistently seeing this issue when trying to obtain bytes from the font template for native fonts on macOS.

Although I am yet to complete the implementation, this is the relevant piece of code that triggers the panic from canvas paint thread:
```rust
let font = with_thread_local_font_context(canvas_data, |font_context| {
    let font_group = font_context.font_group(font_style);
    let font = font_group
        .borrow_mut()
        .first(font_context)
        .expect("couldn't find font");
    let mut font = font.borrow_mut();
    let bytes = font.handle.template().bytes(); // <- panics with 'No URL for Core Text font! (thread CanvasThread, at components/gfx/platform/macos/font_template.rs:143)'
    Font::from_bytes(Arc::new(bytes), 0).unwrap_or_else(load_default_system_fallback_font)
});
```

Also, I noticed we already have [an issue](https://github.com/servo/servo/issues/24622) filed with similar a panic backtrace. I am not sure if it is relevant, but my best guess would be the intermittent test failure can occur when the [font template is accessed](https://github.com/servo/servo/blob/2f6efcbc3a535ed6718457b2f8ab5c78002ca09c/components/gfx/font_cache_thread.rs#L177-L181) before/during the [web font is being fetched asynchronously](https://github.com/servo/servo/blob/2f6efcbc3a535ed6718457b2f8ab5c78002ca09c/components/gfx/font_cache_thread.rs#L200-L202) and eventually hitting:

https://github.com/servo/servo/blob/2f6efcbc3a535ed6718457b2f8ab5c78002ca09c/components/gfx/font_cache_thread.rs#L405

It would be interesting to see if setting the [`load_webfonts_synchronously`](https://github.com/servo/servo/blob/2f6efcbc3a535ed6718457b2f8ab5c78002ca09c/components/config/opts.rs#L70) config option to `true` affects the test failure in any manner.

**Comment by jdm** [comment]
We already run all tests with the synchronous web fonts enabled, unfortunately.

**Comment by utsavoza** [comment]
> We already run all tests with the synchronous web fonts enabled, unfortunately.

Oh, I didn't know that.

In any case, if we consider `FontTemplateData::bytes` a fallible operation, what should be the fallback behavior in case `FontTemplateData::bytes` fails to return the data in the font?

**Comment by jdm** [comment]
These changes look great from a quick skim so far! Do you still need an answer to your last question? I see you're now using bytes_if_in_memory and falling back to loading a system font, which seems reasonable to me. Fantastic effort, in any case!

**Comment by utsavoza** [comment]
> Do you still need an answer to your last question?

Thanks, I just needed to confirm the behavior; nevermind if the current approach seems ok.

I am pasting this here for my reference, will need to rebase the changes and update tests accordingly.

<details>
  <summary>Test results</summary>

```
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: attribute font
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: attribute textAlign
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: attribute textBaseline
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: attribute direction
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "font" with the proper type
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "textAlign" with the proper type
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "textBaseline" with the proper type
▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
└ PASS [expected FAIL] CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "direction" with the proper type

▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.align.default.html:
└ PASS [expected FAIL] Canvas test: 2d.text.align.default
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.align.invalid.html:
└ PASS [expected FAIL] Canvas test: 2d.text.align.invalid
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.baseline.default.html:
└ PASS [expected FAIL] Canvas test: 2d.text.baseline.default
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.font.parse.size.percentage.html:
└ PASS [expected FAIL] Canvas test: 2d.text.font.parse.size.percentage
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.font.default.html:
└ PASS [expected FAIL] Canvas test: 2d.text.font.default
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.baseline.invalid.html:
└ PASS [expected FAIL] Canvas test: 2d.text.baseline.invalid
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.baseline.invalid.html:
└ PASS [expected FAIL] OffscreenCanvas test: 2d.text.baseline.invalid
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.font.parse.size.percentage.default.html:
└ PASS [expected FAIL] Canvas test: 2d.text.font.parse.size.percentage.default
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.align.invalid.worker.html:
└ PASS [expected FAIL] 2d
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.align.default.html:
└ PASS [expected FAIL] OffscreenCanvas test: 2d.text.align.default
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.baseline.default.html:
└ PASS [expected FAIL] OffscreenCanvas test: 2d.text.baseline.default
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.baseline.invalid.worker.html:
└ PASS [expected FAIL] 2d
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.baseline.default.worker.html:
└ PASS [expected FAIL] 2d
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.align.default.worker.html:
└ PASS [expected FAIL] 2d
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.relative_size.worker.html:
└ PASS [expected FAIL] 2d
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.font.parse.system.html:
└ PASS [expected FAIL] System fonts must be computed to explicit values
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.parse.system.html:
└ PASS [expected FAIL] System fonts must be computed to explicit values
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.parse.tiny.html:
│ FAIL [expected PASS] OffscreenCanvas test: 2d.text.font.parse.tiny
│   → assert_equals: ctx.font === '1px sans-serif' (got 10px sans-serif[string], expected 1px sans-serif[string]) expected "1px sans-serif" but got "10px sans-serif"
│
│ _assertSame@http://web-platform.test:8000/html/canvas/resources/canvas-tests.js:17:18
│ @http://web-platform.test:8000/html/canvas/offscreen/text/2d.text.font.parse.tiny.html:24:12
│ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
└ @http://web-platform.test:8000/html/canvas/offscreen/text/2d.text.font.parse.tiny.html:18:3
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.parse.tiny.worker.html:
│ FAIL [expected PASS] 2d
│   → assert_equals: ctx.font === '1px sans-serif' (got 10px sans-serif[string], expected 1px sans-serif[string]) expected "1px sans-serif" but got "10px sans-serif"
│
│ _assertSame@http://web-platform.test:8000/html/canvas/resources/canvas-tests.js:17:18
│ @http://web-platform.test:8000/html/canvas/offscreen/text/2d.text.font.parse.tiny.worker.js:20:12
│ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
└ @http://web-platform.test:8000/html/canvas/offscreen/text/2d.text.font.parse.tiny.worker.js:14:3
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.default.html:
└ PASS [expected FAIL] OffscreenCanvas test: 2d.text.font.default
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.font.relative_size.html:
└ PASS [expected FAIL] Canvas test: 2d.text.font.relative_size
▶ Unexpected subtest result in /html/canvas/element/text-styles/2d.text.font.parse.basic.html:
└ PASS [expected FAIL] Canvas test: 2d.text.font.parse.basic
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.align.invalid.html:
└ PASS [expected FAIL] OffscreenCanvas test: 2d.text.align.invalid
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.default.worker.html:
└ PASS [expected FAIL] 2d
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.parse.system.worker.html:
└ PASS [expected FAIL] System fonts must be computed to explicit values
▶ Unexpected subtest result in /html/canvas/offscreen/text/2d.text.font.relative_size.html:
└ PASS [expected FAIL] OffscreenCanvas test: 2d.text.font.relative_size
▶ Unexpected subtest result in /html/semantics/embedded-content/the-canvas-element/initial.reset.2dstate.html:
└ PASS [expected FAIL] Resetting the canvas state resets 2D state variables
▶ FAIL [expected PASS] /html/canvas/element/text-styles/canvas_text_font_001.htm
│   → /html/canvas/element/text-styles/canvas_text_font_001.htm 54ed3fe6baf2d8bc9bddcdb0e82725a5c568a6e6
└   → /html/canvas/element/text-styles/canvas_text_font_001-ref.htm 271b8adde3002c11379dbc7d18771d3f47b0ed33

▶ Unexpected subtest result in /html/canvas/element/the-canvas-state/2d.state.saverestore.font.html:
└ PASS [expected FAIL] save()/restore() works for font
▶ Unexpected subtest result in /html/canvas/element/the-canvas-state/2d.state.saverestore.textAlign.html:
└ PASS [expected FAIL] save()/restore() works for textAlign
▶ Unexpected subtest result in /html/canvas/element/the-canvas-state/2d.state.saverestore.textBaseline.html:
└ PASS [expected FAIL] save()/restore() works for textBaseline
```
<details>

File: components/style/properties/properties.mako.rs [hunk 1]
```diff
@@ -2601,7 +2601,7 @@ pub mod style_structs {
 
     % for style_struct in data.active_style_structs():
         % if style_struct.name == "Font":
-        #[derive(Clone, Debug, MallocSizeOf)]
+        #[derive(Clone, Debug, MallocSizeOf, Serialize, Deserialize)]
```
Comment (jdm) [inline components/style/properties/properties.mako.rs:2604]: Let's do this:
```rust
#[derive(Clone, Debug, MallocSizeOf)]
#[cfg_attr(feature = "servo", derive(Serialize, Deserialize))]

File: components/style/values/computed/font.rs [hunk 2]
```diff
@@ -80,6 +93,8 @@ impl ToAnimatedValue for FontWeight {
     ToAnimatedZero,
     ToCss,
     ToResolvedValue,
+    Serialize,
+    Deserialize,
```
Comment (jdm) [inline components/style/values/computed/font.rs:97]: Let's make this conditional on the "servo" feature.

File: components/style/values/computed/font.rs [hunk 3]
```diff
@@ -179,7 +194,7 @@ impl ToAnimatedValue for FontSize {
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, ToComputedValue, ToResolvedValue)]
-#[cfg_attr(feature = "servo", derive(Hash, MallocSizeOf))]
+#[cfg_attr(feature = "servo", derive(Hash, MallocSizeOf, Serialize, Deserialize))]
```
Comment (jdm) [inline components/style/values/computed/font.rs:197]: Same here.

File: components/style/values/computed/font.rs [hunk 4]
```diff
@@ -443,9 +458,39 @@ impl SingleFontFamily {
...
+    Debug,
+    Eq,
+    Hash,
+    MallocSizeOf,
+    PartialEq,
+    ToComputedValue,
+    ToResolvedValue,
+    ToShmem,
+    Serialize,
+    Deserialize,
```
Comment (jdm) [inline components/style/values/computed/font.rs:493]: Same here.

File: components/style/values/specified/font.rs [hunk 5]
```diff
@@ -496,6 +496,8 @@ impl ToComputedValue for FontStretch {
     ToCss,
     ToResolvedValue,
     ToShmem,
+    Serialize,
+    Deserialize,
```
Comment (jdm) [inline components/style/values/specified/font.rs:500]: Same here.

File: components/style/values/specified/font.rs [hunk 6]
```diff
@@ -540,6 +542,8 @@ impl Default for KeywordSize {
...
```
Comment (jdm) [inline components/style/values/specified/font.rs:546]: Same here.

**Review by jdm — CHANGES_REQUESTED** [review]
Great job!

File: components/canvas/canvas_data.rs [hunk 7]
```diff
@@ -361,6 +382,21 @@ pub enum Filter {
...
+pub(crate) fn with_thread_local_font_context<F, R>(canvas_data: &CanvasData, f: F) -> R
+where
+    F: FnOnce(&mut CanvasFontContext) -> R,
+{
+    FONT_CONTEXT.with(|font_context| {
+        f(font_context.borrow_mut().get_or_insert_with(|| {
+            FontContext::new(canvas_data.font_cache_thread.lock().unwrap().clone())
+        }))
+    })
+}
```
[the remark this answers is not in the document]
Comment (utsavoza) [reply components/canvas/canvas_data.rs:397]: Actually, we require a mutable reference to CanvasFontContext and storing it CanvasPaintThread and passing a mutable reference in [fillText](https://github.com/servo/servo/pull/26697/files#diff-beef21ae0a52212d45bc2a499bc93eadR498-R505) method might make the borrow checker unhappy.

[the remark this answers is not in the document]
Comment (jdm) [reply components/canvas/canvas_data.rs:397]: I'm not sure I understand why that would be a problem?

[the remark this answers is not in the document]
Comment (utsavoza) [reply components/canvas/canvas_data.rs:397]: I think storing CanvasFontContext and getting a mutable reference will lead to multiple mutable borrows on self [here](https://github.com/servo/servo/pull/26697/files#diff-284fed15fb4a8b4f1991e2c5a9ea1139R148-R149):
```rust            
Canvas2dMsg::FillText(text, x, y, max_width, style, is_rtl) => {
    self.canvas(canvas_id).set_fill_style(style);
    self.canvas(canvas_id)
            .fill_text(text, x, y, max_width, is_rtl);
},

```
as we are calling fill_text on mutably borrowed CanvasData from CanvasPaintThread.

**Pull request merged**